Windows build fails on MinGW GCC 13 due to libssp changes

Compiling Julia on Windows with GCC 13 from MSYS2 (which is the only available compiler) fails during linking:

```
    LINK build2/usr/bin/libjulia-codegen.dll
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ./codegen.o: in function `add_named_global':
C:/Users/Tim/Julia/src/julia/src/codegen.cpp:549:(.text+0x26a53): undefined reference to `__imp___stack_chk_guard'
collect2.exe: error: ld returned 1 exit status
make[1]: *** [/c/Users/Tim/Julia/src/julia/src/Makefile:412: /c/Users/Tim/Julia/src/julia/build2/usr/bin/libjulia-codegen.dll] Error 1
make: *** [/c/Users/Tim/Julia/src/julia/Makefile:97: julia-src-release] Error 2
```

This is on https://github.com/JuliaLang/julia/pull/51698, because otherwise the CSL static libraries break the build.

I'm not entirely sure where we should be getting `__stack_chk_guard` from; it's defined in `rtutils.c`, but `HAVE_SSP` is set, so that code isn't activated:

https://github.com/JuliaLang/julia/blob/4d2d8490bdee8f457948b949441e95a69a775217/src/rtutils.c#L218-L228

Meanwhile, `libssp.dll` from CSL has the symbol:

```
❯ objdump -p ./usr/bin/libssp-0.dll | grep stack
        [   6] __stack_chk_fail
        [   7] __stack_chk_guard
```

... and we do link this DLL, but that somehow doesn't seem to help:

```
g++ -march=native -mtune=native -m64 -shared   -Wl,--out-implib,/c/Users/Tim/Julia/src/julia/build2/usr/lib/libjulia-codegen.dll.a -pipe  -fno-rtti -std=c++14 -fno-gnu-unique   -O3 -ggdb2 -falign-functions -momit-leaf-frame-pointer -D_GNU_SOURCE -I. -I/c/Users/Tim/Julia/src/julia/src -I/c/Users/Tim/Julia/src/julia/src/flisp -I/c/Users/Tim/Julia/src/julia/src/support -I/c/Users/Tim/Julia/src/julia/build2/usr/include -I/c/Users/Tim/Julia/src/julia/build2/usr/include -I/c/Users/Tim/Julia/src/julia/deps/valgrind -Wall -Wno-strict-aliasing -fno-omit-frame-pointer -fvisibility=hidden -fno-common -Wno-comment -Wpointer-arith -Wundef -Wno-unused-result -DJL_BUILD_ARCH='"x86_64"' -DJL_BUILD_UNAME='"NT"' -IC:\Users\Tim\Julia\src\julia\build2\usr\include -DLLVM_SHLIB "-DJL_SYSTEM_IMAGE_PATH=\"../lib/julia/sys.dll\"" "-DJL_LIBJULIA_SONAME=\"libjulia.dll\"" ./codegen.o ./jitlayers.o ./aotcompile.o ./debuginfo.o ./disasm.o ./llvm-simdloop.o ./llvm-muladd.o ./llvm-final-gc-lowering.o ./llvm-pass-helpers.o ./llvm-late-gc-lowering.o ./llvm-ptls.o ./llvm-lower-handlers.o ./llvm-gc-invariant-verifier.o ./llvm-propagate-addrspaces.o ./llvm-multiversioning.o ./llvm-alloc-opt.o ./llvm-alloc-helpers.o ./cgmemmgr.o ./llvm-remove-addrspaces.o ./llvm-remove-ni.o ./llvm-julia-licm.o ./llvm-demote-float16.o ./llvm-cpufeatures.o ./pipeline.o ./llvm_api.o  -o /c/Users/Tim/Julia/src/julia/build2/usr/bin/libjulia-codegen.dll -Wl,--stack,8388608  -L/c/Users/Tim/Julia/src/julia/build2/usr/lib -L/c/Users/Tim/Julia/src/julia/build2/usr/bin  -LC:\Users\Tim\Julia\src\julia\build2\usr/lib  -lLLVM-15jl -Wl,--export-all-symbols -Wl,--version-script=/c/Users/Tim/Julia/src/julia/build2/src/julia.expmap -Wl,--no-whole-archive -lpsapi -lkernel32 -lws2_32 -liphlpapi -lwinmm -ldbghelp -luserenv -lsecur32 -latomic -lssp   -ljulia -ljulia-internal
```

cc @vtjnash

---

See below for a MWE, and the link to upstream changes causing this.

Workaround: `override HAVE_SSP := 0` in `Make.user`.

> The symbol is `__imp____stack_chk_guard` with 4 `_` not 3.

Is it? Even using GCC 12, from our CI's environment, `codegen.o` contains `__imp___stack_chk_guard` with 3 `_`:

```
# objdump -t ../../build/src/codegen.o | grep stack_chk_guard
[158](sec  1)(fl 0x00)(ty   20)(scl   3) (nx 0) 0x0000000000003950 _ZNL21jlstack_chk_guard_varMUlPN4llvm4TypeEE_4_FUNES1_
[650](sec  3)(fl 0x00)(ty    0)(scl   3) (nx 0) 0x0000000000000598 _ZL21jlstack_chk_guard_var
[1614](sec  0)(fl 0x00)(ty    0)(scl   2) (nx 0) 0x0000000000000000 __imp___stack_chk_guard
```

And the `codegen.o` produced by GCC 13 links fine using GCC 12:

```
❯ g++ --version
g++.exe (Rev2, Built by MSYS2 project) 13.2.0
❯ g++ -march=native -mtune=native -m64 -shared   -Wl,--out-implib,/c/Users/Tim/Julia/src/julia/build2/usr/lib/libjulia-codegen.dll.a -pipe -lssp -fno-rtti -std=c++14 -fno-gnu-unique   -O3 -ggdb2 -falign-functions -momit-leaf-frame-pointer -D_GNU_SOURCE -I. -I/c/Users/Tim/Julia/src/julia/src -I/c/Users/Tim/Julia/src/julia/src/flisp -I/c/Users/Tim/Julia/src/julia/src/support -I/c/Users/Tim/Julia/src/julia/build2/usr/include -I/c/Users/Tim/Julia/src/julia/build2/usr/include -I/c/Users/Tim/Julia/src/julia/deps/valgrind -Wall -Wno-strict-aliasing -fno-omit-frame-pointer -fvisibility=hidden -fno-common -Wno-comment -Wpointer-arith -Wundef -Wno-unused-result -DJL_BUILD_ARCH='"x86_64"' -DJL_BUILD_UNAME='"NT"' -IC:\Users\Tim\Julia\src\julia\build2\usr\include -DLLVM_SHLIB "-DJL_SYSTEM_IMAGE_PATH=\"../lib/julia/sys.dll\"" "-DJL_LIBJULIA_SONAME=\"libjulia.dll\"" ./codegen.o ./jitlayers.o ./aotcompile.o ./debuginfo.o ./disasm.o ./llvm-simdloop.o ./llvm-muladd.o ./llvm-final-gc-lowering.o ./llvm-pass-helpers.o ./llvm-late-gc-lowering.o ./llvm-ptls.o ./llvm-lower-handlers.o ./llvm-gc-invariant-verifier.o ./llvm-propagate-addrspaces.o ./llvm-multiversioning.o ./llvm-alloc-opt.o ./llvm-alloc-helpers.o ./cgmemmgr.o ./llvm-remove-addrspaces.o ./llvm-remove-ni.o ./llvm-julia-licm.o ./llvm-demote-float16.o ./llvm-cpufeatures.o ./pipeline.o ./llvm_api.o  -o /c/Users/Tim/Julia/src/julia/build2/usr/bin/libjulia-codegen.dll -Wl,--stack,8388608  -L/c/Users/Tim/Julia/src/julia/build2/usr/lib -L/c/Users/Tim/Julia/src/julia/build2/usr/bin  -LC:\Users\Tim\Julia\src\julia\build2\usr/lib  -lLLVM-15jl -Wl,--export-all-symbols -Wl,--version-script=/c/Users/Tim/Julia/src/julia/build2/src/julia.expmap -Wl,--no-whole-archive -lpsapi -lkernel32 -lws2_32 -liphlpapi -lwinmm -ldbghelp -luserenv -lsecur32 -latomic -ljulia -ljulia-internal
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: ./codegen.o: in function `add_named_global':
C:/Users/Tim/Julia/src/julia/src/codegen.cpp:549:(.text+0x26a53): undefined reference to `__imp___stack_chk_guard'
collect2.exe: error: ld returned 1 exit status

# from within the same folder, i.e., using the same objects

# g++ --version
g++.exe (Rev1, Built by MSYS2 project) 12.2.0
#  g++ -march=native -mtune=native -m64 -shared   -Wl,--out-implib,/c/Users/Tim/Julia/src/julia/build2/usr/lib/libjulia-codegen.dll.a -pipe  -fno-rtti -std=c++14 -fno-gnu-unique   -O3 -ggdb2 -falign-functions -momit-leaf-frame-pointer -D_GNU_SOURCE -I. -I/c/Users/Tim/Julia/src/julia/src -I/c/Users/Tim/Julia/src/julia/src/flisp -I/c/Users/Tim/Julia/src/julia/src/support -I/c/Users/Tim/Julia/src/julia/build2/usr/include -I/c/Users/Tim/Julia/src/julia/build2/usr/include -I/c/Users/Tim/Julia/src/julia/deps/valgrind -Wall -Wno-strict-aliasing -fno-omit-frame-pointer -fvisibility=hidden -fno-common -Wno-comment -Wpointer-arith -Wundef -Wno-unused-result -DJL_BUILD_ARCH='"x86_64"' -DJL_BUILD_UNAME='"NT"' -IC:\Users\Tim\Julia\src\julia\build2\usr\include -DLLVM_SHLIB "-DJL_SYSTEM_IMAGE_PATH=\"../lib/julia/sys.dll\"" "-DJL_LIBJULIA_SONAME=\"libjulia.dll\"" ./codegen.o ./jitlayers.o ./aotcompile.o ./debuginfo.o ./disasm.o ./llvm-simdloop.o ./llvm-muladd.o ./llvm-final-gc-lowering.o ./llvm-pass-helpers.o ./llvm-late-gc-lowering.o ./llvm-ptls.o ./llvm-lower-handlers.o ./llvm-gc-invariant-verifier.o ./llvm-propagate-addrspaces.o ./llvm-multiversioning.o ./llvm-alloc-opt.o ./llvm-alloc-helpers.o ./cgmemmgr.o ./llvm-remove-addrspaces.o ./llvm-remove-ni.o ./llvm-julia-licm.o ./llvm-demote-float16.o ./llvm-cpufeatures.o ./pipeline.o ./llvm_api.o  -o /c/Users/Tim/Julia/src/julia/build2/usr/bin/libjulia-codegen.dll -Wl,--stack,8388608  -L/c/Users/Tim/Julia/src/julia/build2/usr/lib -L/c/Users/Tim/Julia/src/julia/build2/usr/bin  -LC:\Users\Tim\Julia\src\julia\build2\usr/lib  -lLLVM-15jl -Wl,--export-all-symbols -Wl,--version-script=/c/Users/Tim/Julia/src/julia/build2/src/julia.expmap -Wl,--no-whole-archive -lpsapi -lkernel32 -lws2_32 -liphlpapi -lwinmm -ldbghelp -luserenv -lsecur32 -latomic -lssp   -ljulia -ljulia-internal
# success
```

MWE:

```c
extern __declspec(dllimport) int __stack_chk_guard;

int* foo(void)
{
    return &__stack_chk_guard;
}
```
```
❯ gcc -shared test.c -lssp -o test.dll
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccq7LnXo.o:test.c:(.text+0x1b): undefined reference to `__imp___stack_chk_guard'
collect2.exe: error: ld returned 1 exit status
```

Works on GCC 12, fails on 13. Tried some combinations of `-fstack-protector` instead of `-lssp`, to no avail.

--- a/deps/csl.mk
+++ b/deps/csl.mk
@@ -110,7 +110,7 @@ install-csl:
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libgcc_s.a $(build_private_libdir)/
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libgcc.a $(build_private_libdir)/
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libmsvcrt.a $(build_private_libdir)/
-	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libssp.dll.a $(build_private_libdir)/
+	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libssp.dll.a $(build_libdir)/
 endif
 endif
 ifeq ($(OS),WINNT)
@@ -119,5 +119,5 @@ uninstall-gcc-libraries:
 	-rm -f $(build_private_libdir)/libgcc_s.a
 	-rm -f $(build_private_libdir)/libgcc.a
 	-rm -f $(build_private_libdir)/libmsvcrt.a
-	-rm -f $(build_private_libdir)/libssp.dll.a
+	-rm -f $(build_libdir)/libssp.dll.a
 endif

--- a/deps/csl.mk
+++ b/deps/csl.mk
@@ -110,7 +110,8 @@ install-csl:
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libgcc_s.a $(build_private_libdir)/
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libgcc.a $(build_private_libdir)/
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libmsvcrt.a $(build_private_libdir)/
-	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libssp.dll.a $(build_shlibdir)/
+	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libssp.dll.a $(build_private_libdir)/
+	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libssp.dll.a $(build_libdir)/
 endif
 endif
 ifeq ($(OS),WINNT)
@@ -119,5 +120,6 @@ uninstall-gcc-libraries:
 	-rm -f $(build_private_libdir)/libgcc_s.a
 	-rm -f $(build_private_libdir)/libgcc.a
 	-rm -f $(build_private_libdir)/libmsvcrt.a
-	-rm -f $(build_shlibdir)/libssp.dll.a
+	-rm -f $(build_private_libdir)/libssp.dll.a
+	-rm -f $(build_libdir)/libssp.dll.a
 endif

--- a/deps/csl.mk
+++ b/deps/csl.mk
@@ -110,7 +110,7 @@ install-csl:
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libgcc_s.a $(build_private_libdir)/
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libgcc.a $(build_private_libdir)/
 	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libmsvcrt.a $(build_private_libdir)/
-	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libssp.dll.a $(build_libdir)/
+	cp -a $(build_libdir)/gcc/$(BB_TRIPLET)/13/libssp.dll.a $(build_shlibdir)/
 endif
 endif
 ifeq ($(OS),WINNT)
@@ -119,5 +119,5 @@ uninstall-gcc-libraries:
 	-rm -f $(build_private_libdir)/libgcc_s.a
 	-rm -f $(build_private_libdir)/libgcc.a
 	-rm -f $(build_private_libdir)/libmsvcrt.a
-	-rm -f $(build_libdir)/libssp.dll.a
+	-rm -f $(build_shlibdir)/libssp.dll.a
 endif

